### PR TITLE
Make designer and code completion great again

### PR DESCRIPTION
**Pull request by donandren**
- [x]  Update projects to use PackageReferences
- [x]  Fixed Xaml Code completion to work again in general
- [x]  Updated templates to use avalonia 0.7.0
- [x] Fixed broken Xaml type extensions completion

In general avalonia visual studio extension preview and code completion is now usable again:
If anyone want to try it out: https://drive.google.com/open?id=1IVyIcFTUA17LMZ_QFr-_JCGepgyG3pwb

### CODE REVIEW

File: src/AvaloniaVS/Infrastructure/AvaloniaDesignerPane.cs [hunk 1]
```diff
@@ -185,8 +205,15 @@ private async void Restart()
                 return;
             try
             {
-                Console.WriteLine("Restarting designer");
-                _designer?.RestartProcess();
+                if (!IsDesignerVisible())
```
Comment (donandren) [inline src/AvaloniaVS/Infrastructure/AvaloniaDesignerPane.cs]: it was pretty annoying when only xaml code view was active designer to try visualize current control over the xaml.
Writing xaml was almost not possible when designer was reloading .
 Now this changes are preventing this behavior for me, when xaml source only is seen designer is not attempting to reload again and again.

File: src/AvaloniaVS/Infrastructure/AvaloniaPackage.cs [hunk 2]
```diff
@@ -23,7 +23,7 @@ namespace AvaloniaVS.Infrastructure
     // folder
     [ProvideBindingPath]
 
-    [PackageRegistration(UseManagedResourcesOnly = true)]
+    [PackageRegistration(UseManagedResourcesOnly = true, AllowsBackgroundLoading =true)]
```
Comment (donandren) [inline src/AvaloniaVS/Infrastructure/AvaloniaPackage.cs]: some smart analyzer was complaining we need to set AllowsBackgroundLoading to true, not sure it's needed or not

File: src/AvaloniaVS/Infrastructure/PropertyChangedBase.cs [hunk 3]
```diff
@@ -9,7 +9,7 @@
 
 namespace AvaloniaVS.Infrastructure
 {
-    [ImplementPropertyChanged]
+    [AddINotifyPropertyChangedInterface]
```
Comment (donandren) [inline src/AvaloniaVS/Infrastructure/PropertyChangedBase.cs]: ImplementPropertyChanged is obsolete in fody property changed, AddINotifyPropertyChangedInterface should be used instead

Comment (grokys) [reply src/AvaloniaVS/Infrastructure/PropertyChangedBase.cs]: Thanks! I don't use fody so wondered what this was about.

File: src/AvaloniaVS/ProjectProperties.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    using System.Collections.Generic;
+    using System.ComponentModel.Composition;
+    using System.Linq;
+    using System.Text;
+    using System.Threading.Tasks;
+    using Microsoft.VisualStudio.ProjectSystem;
+    using Microsoft.VisualStudio.ProjectSystem.Properties;
+
+    [Export]
+    internal partial class ProjectProperties : StronglyTypedPropertyAccess
```
Comment (donandren) [inline src/AvaloniaVS/ProjectProperties.cs]: this is some auto added file when packages was updated, no idea it's needed or not

File: src/AvaloniaVS/Infrastructure/AvaloniaPackage.cs (hunk 2 above)
Comment (grokys) [reply src/AvaloniaVS/Infrastructure/AvaloniaPackage.cs]: Yep, this one I understand, and it's a good thing to have here :D

File: src/AvaloniaVS/ProjectProperties.cs (hunk 4 above)
Comment (grokys) [reply src/AvaloniaVS/ProjectProperties.cs]: Hmm yeah this is strange. I've never seen `StronglyTypedPropertyAccess` before and can't see any documentation on it?

**Comment by grokys** [comment]
> NewtonSoft.Json reference is downgraded to 9.0.0.0, noticed some bad vs exceptions :( when latter version is used - quite strange behavior i don't understand and can't find logic why

Yeah, VS itself ships 9.0.1 and has binding redirects to this version from lower versions. It's strongly discouraged to use a version higher than this in VS because it causes breakage: https://github.com/github/VisualStudio/issues/1774

**Comment by grokys** [comment]
@donandren I'm unable to sync the submodule, getting:

```
> git submodule update
error: Server does not allow request for unadvertised object 351feb0d1b15d0c96674706b3cc9b68cbb81e9e1
Fetched in submodule path 'src/external/Avalonia.Ide', but it did not contain 351feb0d1b15d0c96674706b3cc9b68cbb81e9e1. Direct fetching of that commit failed.
```

This is on a fresh clone. Do you need to push a submodule update?

File: src/AvaloniaVS/ProjectProperties.cs (hunk 4 above)
Comment (jp2masa) [reply src/AvaloniaVS/ProjectProperties.cs]: That is CPS. It was added by the `Microsoft.VisualStudio.ProjectSystem.Sdk` package. As you're not writing a project system, it's not needed I think.

**Comment by donandren** [comment]
this is the submodule pushed -https://github.com/donandren/Avalonia.Ide/commit/351feb0d1b15d0c96674706b3cc9b68cbb81e9e1

but may be better Avalonia.Ide submodule be under AvaloniaUI

**Comment by donandren** [comment]
@grokys i haven't reformatted the code on purpose so changes are much easy to see, but all code of AvaloniaVS need reformat, so one is not constantly reverting automatic formatting changes 😄 
if you can reformat after eventual merge would be nice

File: src/AvaloniaVS/Infrastructure/PropertyChangedBase.cs (hunk 3 above)
Comment (worldbeater) [reply src/AvaloniaVS/Infrastructure/PropertyChangedBase.cs]: There is no need to use the `AddINotifyPropertyChangedInterface` attribute here, see [PropertyChanged.Fody wiki](https://github.com/Fody/PropertyChanged/wiki/Attributes#addinotifypropertychangedinterfaceattribute). Worth removing either the attribute or the `INotifyPropertyChanged` interface implementation.

> AddINotifyPropertyChangedInterfaceAttribute. Specifies that the class will be marked with INotifyPropertyChanged. Note that all classes that implement INotifyPropertyChanged will have property notification injected irrespective of the use of this attribute.

**Comment by grokys** [comment]
Yeah feel free to submit a reformatting PR after this is merged!

File: src/AvaloniaVS/ProjectProperties.cs (hunk 4 above)
Comment (grokys) [reply src/AvaloniaVS/ProjectProperties.cs]: Thanks @jp2masa - this should be removed I think.

**Comment by donandren** [comment]
> I'm now getting a load of build errors saying:
> 
> ```
> 3>d:\projects\AvaloniaUI\AvaloniaVS\src\AvaloniaVS.Controls\Internals\AvaloniaDesignerBrushes.cs(5,17,5,29): error CS0234: The type or namespace name 'VisualStudio' does not exist in the namespace 'Microsoft' (are you missing an assembly reference?)
> 3>d:\projects\AvaloniaUI\AvaloniaVS\src\AvaloniaVS.Controls\Internals\AvaloniaDesignerBrushes.cs(19,23,19,39): error CS0246: The type or namespace name 'ThemeResourceKey' could not be found (are you missing a using directive or an assembly reference?)
> ```
> Which is strange because `Microsoft.VisualStudio.Shell.Immutable.11.0` should be including that. Is anyone else seeing the same?

this error has been addressed here https://github.com/AvaloniaUI/AvaloniaVS/pull/76/commits/8602e49242bf48901faa165e6177055d6b973372

clean check out, git clean or delete all bin/obj/old packages folders then nuget restore should fix it

**Comment by grokys** [comment]
I notice you're still working on this ;) Do let me know when it's ready for review/merge!

**Comment by donandren** [comment]
@grokys it's ready, not sure i'll spend more time, as now it's good enough for me

**Comment by grokys** [comment]
I'm still getting build errors:

```
NU1603: Microsoft.VisualStudio.Language.Intellisense 15.8.525 depends on Microsoft.VisualStudio.ImageCatalog (>= 15.8.27731) but Microsoft.VisualStudio.ImageCatalog 15.8.27731 was not found. An approximate best match of Microsoft.VisualStudio.ImageCatalog 15.8.28010 was resolved.
NU1603: Microsoft.VisualStudio.Shell.Immutable.14.0 15.0.25405 depends on Microsoft.VisualStudio.Shell.Framework (>= 15.0.25405) but Microsoft.VisualStudio.Shell.Framework 15.0.25405 was not found. An approximate best match of Microsoft.VisualStudio.Shell.Framework 15.0.26201 was resolved.
NU1603: Microsoft.VisualStudio.Language 15.8.525 depends on StreamJsonRpc (>= 1.3.6) but StreamJsonRpc 1.3.6 was not found. An approximate best match of StreamJsonRpc 1.3.23 was resolved.
1>------ Build started: Project: AvaloniaApplicationTemplate, Configuration: Debug Any CPU ------
2>------ Build started: Project: AvaloniaWindowTemplate, Configuration: Debug Any CPU ------
3>------ Build started: Project: AvaloniaVS.Controls, Configuration: Debug Any CPU ------
4>------ Build started: Project: AvaloniaUserControlTemplate, Configuration: Debug Any CPU ------
5>------ Build started: Project: AvaloniaMvvmApplicationTemplate, Configuration: Debug Any CPU ------
3>C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(2110,5): warning MSB3277: Found conflicts between different versions of "Microsoft.VisualStudio.Threading" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed.
3>C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Microsoft.Common.CurrentVersion.targets(2110,5): warning MSB3277: Found conflicts between different versions of "Newtonsoft.Json" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed.
3>  AvaloniaVS.Controls -> D:\projects\AvaloniaUI\AvaloniaVS\src\AvaloniaVS.Controls\bin\Debug\AvaloniaVS.Controls.dll
6>------ Build started: Project: AvaloniaVS, Configuration: Debug Any CPU ------
6>C:\Users\groky\.nuget\packages\microsoft.vssdk.buildtools\15.8.3252\tools\VSSDK\Microsoft.VsSDK.targets(83,5): warning VSSDK1018: Targets version (1.0.612.51260) and build task version (1.0.613.19362) do not match.  Restarting Visual Studio could help fix this issue.
6>C:\Users\groky\.nuget\packages\microsoft.vssdk.buildtools\15.8.3252\tools\VSSDK\Microsoft.VsSDK.targets(87,5): warning VSSDK1018: Targets version (15.8.3252.51260) and build task version (15.8.3253.19362) do not match.  Restarting Visual Studio could help fix this issue.
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine\obj\Debug\netstandard1.4\Avalonia.Ide.CompletionEngine.AssemblyInfo.cs(14,12,14,54): error CS0579: Duplicate 'System.Reflection.AssemblyCompanyAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine\obj\Debug\netstandard1.4\Avalonia.Ide.CompletionEngine.AssemblyInfo.cs(15,12,15,60): error CS0579: Duplicate 'System.Reflection.AssemblyConfigurationAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine\obj\Debug\netstandard1.4\Avalonia.Ide.CompletionEngine.AssemblyInfo.cs(16,12,16,58): error CS0579: Duplicate 'System.Reflection.AssemblyFileVersionAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine\obj\Debug\netstandard1.4\Avalonia.Ide.CompletionEngine.AssemblyInfo.cs(18,12,18,54): error CS0579: Duplicate 'System.Reflection.AssemblyProductAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine\obj\Debug\netstandard1.4\Avalonia.Ide.CompletionEngine.AssemblyInfo.cs(19,12,19,52): error CS0579: Duplicate 'System.Reflection.AssemblyTitleAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine\obj\Debug\netstandard1.4\Avalonia.Ide.CompletionEngine.AssemblyInfo.cs(20,12,20,54): error CS0579: Duplicate 'System.Reflection.AssemblyVersionAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine\obj\Release\netstandard1.4\Avalonia.Ide.CompletionEngine.AssemblyInfo.cs(14,12,14,54): error CS0579: Duplicate 'System.Reflection.AssemblyCompanyAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine\obj\Release\netstandard1.4\Avalonia.Ide.CompletionEngine.AssemblyInfo.cs(15,12,15,60): error CS0579: Duplicate 'System.Reflection.AssemblyConfigurationAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine\obj\Release\netstandard1.4\Avalonia.Ide.CompletionEngine.AssemblyInfo.cs(16,12,16,58): error CS0579: Duplicate 'System.Reflection.AssemblyFileVersionAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine\obj\Release\netstandard1.4\Avalonia.Ide.CompletionEngine.AssemblyInfo.cs(17,12,17,67): error CS0579: Duplicate 'System.Reflection.AssemblyInformationalVersionAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine\obj\Release\netstandard1.4\Avalonia.Ide.CompletionEngine.AssemblyInfo.cs(18,12,18,54): error CS0579: Duplicate 'System.Reflection.AssemblyProductAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine\obj\Release\netstandard1.4\Avalonia.Ide.CompletionEngine.AssemblyInfo.cs(19,12,19,52): error CS0579: Duplicate 'System.Reflection.AssemblyTitleAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine\obj\Release\netstandard1.4\Avalonia.Ide.CompletionEngine.AssemblyInfo.cs(20,12,20,54): error CS0579: Duplicate 'System.Reflection.AssemblyVersionAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine.SrmMetadataProvider\obj\Debug\netstandard1.4\Avalonia.Ide.CompletionEngine.SrmMetadataProvider.AssemblyInfo.cs(14,12,14,54): error CS0579: Duplicate 'System.Reflection.AssemblyCompanyAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine.SrmMetadataProvider\obj\Debug\netstandard1.4\Avalonia.Ide.CompletionEngine.SrmMetadataProvider.AssemblyInfo.cs(15,12,15,60): error CS0579: Duplicate 'System.Reflection.AssemblyConfigurationAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine.SrmMetadataProvider\obj\Debug\netstandard1.4\Avalonia.Ide.CompletionEngine.SrmMetadataProvider.AssemblyInfo.cs(16,12,16,58): error CS0579: Duplicate 'System.Reflection.AssemblyFileVersionAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine.SrmMetadataProvider\obj\Debug\netstandard1.4\Avalonia.Ide.CompletionEngine.SrmMetadataProvider.AssemblyInfo.cs(17,12,17,67): error CS0579: Duplicate 'System.Reflection.AssemblyInformationalVersionAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine.SrmMetadataProvider\obj\Debug\netstandard1.4\Avalonia.Ide.CompletionEngine.SrmMetadataProvider.AssemblyInfo.cs(18,12,18,54): error CS0579: Duplicate 'System.Reflection.AssemblyProductAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine.SrmMetadataProvider\obj\Debug\netstandard1.4\Avalonia.Ide.CompletionEngine.SrmMetadataProvider.AssemblyInfo.cs(19,12,19,52): error CS0579: Duplicate 'System.Reflection.AssemblyTitleAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine.SrmMetadataProvider\obj\Debug\netstandard1.4\Avalonia.Ide.CompletionEngine.SrmMetadataProvider.AssemblyInfo.cs(20,12,20,54): error CS0579: Duplicate 'System.Reflection.AssemblyVersionAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine.SrmMetadataProvider\obj\Release\netstandard1.4\Avalonia.Ide.CompletionEngine.SrmMetadataProvider.AssemblyInfo.cs(14,12,14,54): error CS0579: Duplicate 'System.Reflection.AssemblyCompanyAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine.SrmMetadataProvider\obj\Release\netstandard1.4\Avalonia.Ide.CompletionEngine.SrmMetadataProvider.AssemblyInfo.cs(15,12,15,60): error CS0579: Duplicate 'System.Reflection.AssemblyConfigurationAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine.SrmMetadataProvider\obj\Release\netstandard1.4\Avalonia.Ide.CompletionEngine.SrmMetadataProvider.AssemblyInfo.cs(16,12,16,58): error CS0579: Duplicate 'System.Reflection.AssemblyFileVersionAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine.SrmMetadataProvider\obj\Release\netstandard1.4\Avalonia.Ide.CompletionEngine.SrmMetadataProvider.AssemblyInfo.cs(17,12,17,67): error CS0579: Duplicate 'System.Reflection.AssemblyInformationalVersionAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine.SrmMetadataProvider\obj\Release\netstandard1.4\Avalonia.Ide.CompletionEngine.SrmMetadataProvider.AssemblyInfo.cs(18,12,18,54): error CS0579: Duplicate 'System.Reflection.AssemblyProductAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine.SrmMetadataProvider\obj\Release\netstandard1.4\Avalonia.Ide.CompletionEngine.SrmMetadataProvider.AssemblyInfo.cs(19,12,19,52): error CS0579: Duplicate 'System.Reflection.AssemblyTitleAttribute' attribute
6>D:\projects\AvaloniaUI\AvaloniaVS\src\external\Avalonia.Ide\src\Avalonia.Ide.CompletionEngine.SrmMetadataProvider\obj\Release\netstandard1.4\Avalonia.Ide.CompletionEngine.SrmMetadataProvider.AssemblyInfo.cs(20,12,20,54): error CS0579: Duplicate 'System.Reflection.AssemblyVersionAttribute' attribute
========== Build: 5 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

**Comment by grokys** [comment]
I think we need to set up CI on this repo! I'll look into doing that now.

**Comment by grokys** [comment]
Merged! Thanks @donandren. We should probably get this uploaded to the VS marketplace!

**Comment by TheColonel2688** [comment]
This still hasn't been uploaded to VS Marketplace as far as I can tell.

**Comment by grokys** [comment]
@TheColonel2688 unfortunately not - there are still serious problem with it.

I've been working on a rewrite over at https://github.com/grokys/AvaloniaVS which should fix many of the problems the current version has (it uses a completely different previewer engine) but it's not _quite_ ready yet.

I hope to have it ready for the 0.8 release which should be coming very soon. In the meantime are you able to build the version over at https://github.com/grokys/AvaloniaVS and give it a try? If not, let me know and I will upload a build somewhere.

**Comment by TheColonel2688** [comment]
I should be able to give it a try, tomorrow or sunday.

**Pull request merged**